libc/src/errno/libc_errno.cpp

```diff
@@ -9,44 +9,91 @@
+#define LIBC_ERRNO_MODE_UNDEFINED 0x01
+#define LIBC_ERRNO_MODE_THREAD_LOCAL 0x02
+#define LIBC_ERRNO_MODE_GLOBAL 0x04
+#define LIBC_ERRNO_MODE_EXTERNAL 0x08
+#define LIBC_ERRNO_MODE_SYSTEM 0x10
+
+#ifndef LIBC_ERRNO_MODE
+#ifndef LIBC_COPT_PUBLIC_PACKAGING
+// This mode is for unit testing. We just use our internal errno.
+#define LIBC_ERRNO_MODE LIBC_ERRNO_MODE_INTERNAL
+#elif defined(LIBC_FULL_BUILD)
+// In full build mode, we provide the errno storage ourselves.
+#define LIBC_ERRNO_MODE LIBC_ERRNO_MODE_THREAD_LOCAL
+#else
+#define LIBC_ERRNO_MODE LIBC_ERRNO_MODE_SYSTEM
+#endif
+#endif // LIBC_ERRNO_MODE
+
+#if LIBC_ERRNO_MODE != LIBC_ERRNO_MODE_UNDEFINED &&                            \
+    LIBC_ERRNO_MODE != LIBC_ERRNO_MODE_THREAD_LOCAL &&                         \
+    LIBC_ERRNO_MODE != LIBC_ERRNO_MODE_GLOBAL &&                               \
+    LIBC_ERRNO_MODE != LIBC_ERRNO_MODE_EXTERNAL &&                             \
+    LIBC_ERRNO_MODE != LIBC_ERRNO_MODE_SYSTEM
+#error LIBC_ERRNO_MODE must be one of the following values: \
+LIBC_ERRNO_MODE_NONE, \
+LIBC_ERRNO_MODE_THREAD_LOCAL, \
+LIBC_ERRNO_MODE_GLOBAL, \
+LIBC_ERRNO_MODE_EXTERNAL, \
+LIBC_ERRNO_MODE_SYSTEM
+#endif
+
+namespace LIBC_NAMESPACE {
+
+// Define the global `libc_errno` instance.
+Errno libc_errno;
+
+#if LIBC_ERRNO_MODE == LIBC_ERRNO_MODE_UNDEFINED
+
+void Errno::operator=(int) {}
+Errno::operator int() { return 0; }
+
+#elif LIBC_ERRNO_MODE == LIBC_ERRNO_MODE_THREAD_LOCAL
+
+namespace {
+LIBC_THREAD_LOCAL int __libc_errno;
+}
+
 extern "C" {
-LIBC_THREAD_LOCAL LIBC_NAMESPACE::cpp::Atomic<int> __llvmlibc_errno;
+int *__llvm_libc_errno(void) { return &__libc_errno; }
 }
 
-void LIBC_NAMESPACE::Errno::operator=(int a) {
-  __llvmlibc_errno.store(a, cpp::MemoryOrder::RELAXED);
+void Errno::operator=(int a) { __libc_errno = a; }
+Errno::operator int() { return __libc_errno; }
+
+#elif LIBC_ERRNO_MODE == LIBC_ERRNO_MODE_GLOBAL
+
+namespace {
+LIBC_NAMESPACE::cpp::Atomic<int> __libc_errno;
 }
-LIBC_NAMESPACE::Errno::operator int() {
-  return __llvmlibc_errno.load(cpp::MemoryOrder::RELAXED);
+
+extern "C" {
+int *__llvm_libc_errno(void) { return &__libc_errno; }
 }
 
-#elif !defined(LIBC_COPT_PUBLIC_PACKAGING)
-// This mode is for unit testing.  We just use our internal errno.
-LIBC_THREAD_LOCAL int __llvmlibc_internal_errno;
+void Errno::operator=(int a) {
+  __libc_errno.store(a, cpp::MemoryOrder::RELAXED);
+}
+Errno::operator int() {
+  return __libc_errno.load(cpp::MemoryOrder::RELAXED);
+}
 
-void LIBC_NAMESPACE::Errno::operator=(int a) { __llvmlibc_internal_errno = a; }
-LIBC_NAMESPACE::Errno::operator int() { return __llvmlibc_internal_errno; }
+#elif LIBC_ERRNO_MODE == LIBC_ERRNO_MODE_EXTERNAL
 
-#elif defined(LIBC_FULL_BUILD)
-// This mode is for public libc archive, hermetic, and integration tests.
-// In full build mode, we provide the errno storage ourselves.
 extern "C" {
-LIBC_THREAD_LOCAL int __llvmlibc_errno;
+int *__llvm_libc_errno(void);
```

Comment:
You don't need a declaration here. It's already in `<errno.h>`.

Reply:
We don't include `errno.h` in the full build mode (see https://github.com/llvm/llvm-project/blob/ccceeeb577878f3e9b7301042f68e37552a3d2c1/libc/hdr/errno_macros.h), should we?

Reply:
I think there's a policy about using a `hdr/errno_macros.h` wrapper instead of doing it directly, but yeah.  In some fashion, the authoritative header file declaration of any public ABI function you're defining really ought to be in scope in the implementation file.

Reply:
I think there may be a case to be made that there should be an `__llvm_libc_errno.h` in this directory to declare the function like we have for other public (ABI) functions.  But then that would go with the `__llvm_libc_errno.cpp` implementation file for just that function, which isn't what we want to do since we want it and the `Errno` internal class implemented together.

I'm not sure what the precedent is (or if there is any yet) for other non-functions in the API, i.e. having an `errno.h` that's responsible for `errno` inside the libc implementation.

Reply:
> I'm not sure what the precedent is (or if there is any yet) for other non-functions in the API, i.e. having an `errno.h` that's responsible for `errno` inside the libc implementation.

We use https://github.com/llvm/llvm-project/blob/0b15f89182a4b2a4c46ad207fa2e282ad35f12ee/libc/src/errno/libc_errno.h for that.

Reply:
That header has that name because it's implementing `LIBC_NAMESPACE::libc_errno`.  It's true that today it does include `hdr/errno_macros.h`, but that only defines the `E*` constants (necessarily, i.e. in all build modes).

That's a different issue from whether when there is a variable (or variable-like API, such as `errno`) in the public API, we have a `symbol_name.h` header for it in `src/something/` like we do for functions in the public API.

Reply:
The only other case I can think of is https://github.com/llvm/llvm-project/blob/7868033d2e846fa30c20455ca819fad29d9d795e/libc/src/assert/assert.h which would suggest having `errno.h` in https://github.com/llvm/llvm-project/tree/7868033d2e846fa30c20455ca819fad29d9d795e/libc/src/errno

Reply:
I introduced `src/errno/errno.h` in the latest revision.